docs/pattern-matching.md

```diff
@@ -1,229 +1,306 @@
+```reason
+switch (x) {
+| 0 => "zero"
+| _ => "another number"
+};
+```
+
+Using catch-all variable like `_` is useful when handling many values, such as with ints and strings. But if there are fewer possibilities, such as with variants, it's usually better to match each explicitly to ensure that all cases are handled and guard against future changes.
 
-Do not abuse the fall-through `_` case too much. This prevents the compiler from telling you that you've forgotten to cover a case (exhaustiveness check), which would be especially helpful after a refactoring where you add a new case to a variant. Try only using `_` against infinite possibilities, e.g. string, int, etc.
+### Unused warning
 
-Here's a series of examples, from worst to best:
+The compiler also returns a warning if patterns are repeated.
 
 ```reason
-let optionBoolToBool = opt =>
-  if (opt == None) {
-    false;
-  } else if (opt == Some(true)) {
-    true;
-  } else {
-    false;
-  };
+let x = 3;
+switch (x) {
+| 0 => "zero"
+| 0 => "nil" /* Warning: this match case is unused. */
+| k => "another number " ++ string_of_int(k)
+};
 ```
 
-Now that's just silly =). Let's turn it into pattern-matching:
+More generally, this warning detects when a pattern will never be matched due to the patterns above it.
 
 ```reason
-let optionBoolToBool = opt =>
-  switch (opt) {
-  | None => false
-  | Some(a) => a ? true : false
-  };
+switch (x) {
+| k => "another number " ++ string_of_int(k)
+| 0 => "zero" /* Warning: this match case is unused. */
+};
 ```
 
-Slightly better, but still nested. Pattern-matching allows you to do this:
+Since patterns are matched sequentially, this warning can sometimes be resolved by changing the order in which they're listed. 
 
 ```reason
-let optionBoolToBool = opt =>
-  switch (opt) {
-  | None => false
-  | Some(true) => true
-  | Some(false) => false
-  };
+/* no warning */
+switch (x) {
+| 0 => "zero" 
+| k => "another number " ++ string_of_int(k)
+};
 ```
 
-Much more linear-looking! Now, you might be tempted to do this:
+## Using patterns outside of switch statements
+
+Patterns can also be used outside of switch statements to "unpack" data whenever variables are declared.
 
 ```reason
-let optionBoolToBool = opt =>
-  switch (opt) {
-  | Some(true) => true
-  | _ => false
-  };
+let data = (1, ("red", true));
+let (a, (b, _) as c) = data;
+/* a is 1, b is "red", c is ("red", true) */
+
+type point = {
+  x: int,
+  y: int,
+};
+let f = ({x, y} as p) => x + y + p.x + p.y;
 ```
 
-Which is much more concise, but kills the exhaustiveness check mentioned above; refrain from using that. This is the best:
+Declarations can combine patterns with `|` if the same variables are declared in each pattern.
 
 ```reason
-let optionBoolToBool = opt =>
-  switch (opt) {
-  | Some(trueOrFalse) => trueOrFalse
-  | None => false
-  };
+type t = A(int) | B(int);
+let data = B(3);
+let A(v) | B(v) = data; 
```

Comment:
I think this is probable more confusing than helpful, maybe it makes sense to leave this out?